results/Log_11011055/parameters.txt
# -----------------------------------#
# Parameters of the training session #
# -----------------------------------#

# Input parameters
# ****************

dataset = 3DMatch
network_model = descriptor
num_classes = 0
in_points_dim = 3
in_features_dim = 1
in_radius = 1.000
input_threads = 8

# Model parameters
# ****************

architecture = simple resnetb resnetb_strided resnetb resnetb_strided resnetb resnetb_strided resnetb resnetb_strided resnetb nearest_upsample unary nearest_upsample unary nearest_upsample unary nearest_upsample unary last_unary
num_layers = 5
first_features_dim = 64
use_batch_norm = 1
batch_norm_momentum = 0.980

segmentation_ratio = 1.000

# KPConv parameters
# *****************

first_subsampling_dl = 0.030
num_kernel_points = 15
density_parameter = 5.000
fixed_kernel_points = center
KP_extent = 1.000
KP_influence = linear
convolution_mode = sum
trainable_positions = 0

modulated = 0

# Training parameters
# *******************

learning_rate = 0.050000
momentum = 0.980000
lr_decay_epochs = 1:0.971628 2:0.971628 3:0.971628 4:0.971628 5:0.971628 6:0.971628 7:0.971628 8:0.971628 9:0.971628 10:0.971628 11:0.971628 12:0.971628 13:0.971628 14:0.971628 15:0.971628 16:0.971628 17:0.971628 18:0.971628 19:0.971628 20:0.971628 21:0.971628 22:0.971628 23:0.971628 24:0.971628 25:0.971628 26:0.971628 27:0.971628 28:0.971628 29:0.971628 30:0.971628 31:0.971628 32:0.971628 33:0.971628 34:0.971628 35:0.971628 36:0.971628 37:0.971628 38:0.971628 39:0.971628 40:0.971628 41:0.971628 42:0.971628 43:0.971628 44:0.971628 45:0.971628 46:0.971628 47:0.971628 48:0.971628 49:0.971628 50:0.971628 51:0.971628 52:0.971628 53:0.971628 54:0.971628 55:0.971628 56:0.971628 57:0.971628 58:0.971628 59:0.971628 60:0.971628 61:0.971628 62:0.971628 63:0.971628 64:0.971628 65:0.971628 66:0.971628 67:0.971628 68:0.971628 69:0.971628 70:0.971628 71:0.971628 72:0.971628 73:0.971628 74:0.971628 75:0.971628 76:0.971628 77:0.971628 78:0.971628 79:0.971628 80:0.971628 81:0.971628 82:0.971628 83:0.971628 84:0.971628 85:0.971628 86:0.971628 87:0.971628 88:0.971628 89:0.971628 90:0.971628 91:0.971628 92:0.971628 93:0.971628 94:0.971628 95:0.971628 96:0.971628 97:0.971628 98:0.971628 99:0.971628 100:0.971628 101:0.971628 102:0.971628 103:0.971628 104:0.971628 105:0.971628 106:0.971628 107:0.971628 108:0.971628 109:0.971628 110:0.971628 111:0.971628 112:0.971628 113:0.971628 114:0.971628 115:0.971628 116:0.971628 117:0.971628 118:0.971628 119:0.971628 120:0.971628 121:0.971628 122:0.971628 123:0.971628 124:0.971628 125:0.971628 126:0.971628 127:0.971628 128:0.971628 129:0.971628 130:0.971628 131:0.971628 132:0.971628 133:0.971628 134:0.971628 135:0.971628 136:0.971628 137:0.971628 138:0.971628 139:0.971628 140:0.971628 141:0.971628 142:0.971628 143:0.971628 144:0.971628 145:0.971628 146:0.971628 147:0.971628 148:0.971628 149:0.971628 150:0.971628 151:0.971628 152:0.971628 153:0.971628 154:0.971628 155:0.971628 156:0.971628 157:0.971628 158:0.971628 159:0.971628 160:0.971628 161:0.971628 162:0.971628 163:0.971628 164:0.971628 165:0.971628 166:0.971628 167:0.971628 168:0.971628 169:0.971628 170:0.971628 171:0.971628 172:0.971628 173:0.971628 174:0.971628 175:0.971628 176:0.971628 177:0.971628 178:0.971628 179:0.971628 180:0.971628 181:0.971628 182:0.971628 183:0.971628 184:0.971628 185:0.971628 186:0.971628 187:0.971628 188:0.971628 189:0.971628 190:0.971628 191:0.971628 192:0.971628 193:0.971628 194:0.971628 195:0.971628 196:0.971628 197:0.971628 198:0.971628 199:0.971628
grad_clip_norm = 100.000000

augment_symmetries = 0 0 0
augment_rotation = 1
augment_noise = 0.005000
augment_occlusion = none
augment_occlusion_ratio = 0.200
augment_occlusion_num = 1
augment_scale_anisotropic = 1
augment_scale_min = 0.900
augment_scale_max = 1.100
augment_color = 0.700

weights_decay = 0.000001
gaussian_decay = 0.001000
batch_averaged_loss = 0
offsets_loss = fitting
offsets_decay = 0.100000
batch_num = 1
max_epoch = 200
epoch_steps = 2918
validation_size = 500
snapshot_gap = 1


utils/config.py
#
#
#      0=================================0
#      |    Kernel Point Convolutions    |
#      0=================================0
#
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Configuration class
#
# ----------------------------------------------------------------------------------------------------------------------
#
#      Hugues THOMAS - 11/06/2018
#


from os.path import join


class Config:
    """
    Class containing the parameters you want to modify for this dataset
    """

    ##################
    # Input parameters
    ##################
    # parameter added.
    gpu_id = 0
    keypts_num = 16
    repeat_loss_weight = 0
    safe_radius = 0.10

    # Dataset name
    dataset = ''

    # Type of network model
    network_model = ''

    # Number of classes in the dataset
    num_classes = 0

    # Dimension of input points
    in_points_dim = 3

    # Dimension of input features
    in_features_dim = 1

    # Radius of the input sphere (ignored for models, only used for point clouds)
    in_radius = 1.0

    # Number of CPU threads for the input pipeline
    input_threads = 8

    ##################
    # Model parameters
    ##################

    # Architecture definition. List of blocks
    architecture = []

    # Dimension of the first feature maps
    first_features_dim = 64

    # Batch normalization parameters
    use_batch_norm = True
    batch_norm_momentum = 0.99

    # For segmentation models : ratio between the segmented area and the input area
    segmentation_ratio = 1.0

    ###################
    # KPConv parameters
    ###################

    # First size of grid used for subsampling
    first_subsampling_dl = 0.02

    # Radius of the kernels in the first layer (deprecated)
    first_kernel_radius = 0.1

    # Number of points in the kernels
    num_kernel_points = 15

    # density of neighbors in kernel range
    # For each layer, support points are subsampled on a grid with dl = kernel_radius / density_parameter
    density_parameter = 3.0

    # Kernel point influence radius
    KP_extent = 1.0

    # Influence function when d < KP_extent. ('constant', 'linear', 'gaussian') When d > KP_extent, always zero
    KP_influence = 'gaussian'

    # Behavior of convolutions in ('closest', 'sum')
    # Decide if you sum all kernel point influences, or if you only take the influence of the closest KP
    convolution_mode = 'closest'

    # Fixed points in the kernel : 'none', 'center' or 'verticals'
    fixed_kernel_points = 'center'

    # Can the network learn kernel dispositions (deprecated)
    trainable_positions = False

    # Use modulateion in deformable convolutions
    modulated = False

    #####################
    # Training parameters
    #####################

    # Network optimizer parameters (learning rate and momentum)
    learning_rate = 1e-4
    momentum = 0.9

    # Learning rate decays. Dictionary of all decay values with their epoch {epoch: decay}.
    lr_decays = {200: 0.2, 300: 0.2}

    # Gradient clipping value (negative means no clipping)
    grad_clip_norm = 100.0

    # Augmentation parameters
    augment_scale_anisotropic = True
    augment_scale_min = 0.9
    augment_scale_max = 1.1
    augment_symmetries = [False, False, False]
    augment_rotation = 'vertical'
    augment_noise = 0.005
    augment_occlusion = 'planar'
    augment_occlusion_ratio = 0.2
    augment_occlusion_num = 1
    augment_color = 0.7
    augment_shift_range = 0

    # Regularization loss importance
    weights_decay = 1e-6

    # Gaussian loss
    gaussian_decay = 1e-3

    # Type of output loss with regard to batches when segmentation
    batch_averaged_loss = False

    # Point loss DPRECATED
    points_loss = ''
    points_decay = 1e-2

    # Offset regularization loss
    offsets_loss = 'permissive'
    offsets_decay = 1e-2

    # Number of batch
    batch_num = 10

    # Maximal number of epochs
    max_epoch = 1000

    # Number of steps per epochs
    epoch_steps = 1000

    # Number of validation examples per epoch
    validation_size = 100

    # Number of epoch between each snapshot
    snapshot_gap = 50

    # Do we nee to save convergence
    saving = True
    saving_path = None

    def __init__(self):
        """
        Class Initialyser
        """

        # Number of layers
        self.num_layers = len([block for block in self.architecture if 'pool' in block or 'strided' in block]) + 1

    def load(self, path):

        filename = join(path, 'parameters.txt')
        with open(filename, 'r') as f:
            lines = f.readlines()

        # Class variable dictionary
        for line in lines:
            line_info = line.split()
            if len(line_info) > 1 and line_info[0] != '#':

                if line_info[2] == 'None':
                    setattr(self, line_info[0], None)

                elif line_info[0] == 'lr_decay_epochs':
                    self.lr_decays = {int(b.split(':')[0]): float(b.split(':')[1]) for b in line_info[2:]}

                elif line_info[0] == 'architecture':
                    self.architecture = [b for b in line_info[2:]]

                elif line_info[0] == 'augment_symmetries':
                    self.augment_symmetries = [bool(int(b)) for b in line_info[2:]]

                elif line_info[0] == 'num_classes':
                    if len(line_info) > 3:
                        self.num_classes = [int(c) for c in line_info[2:]]
                    else:
                        self.num_classes = int(line_info[2])

                else:

                    attr_type = type(getattr(self, line_info[0]))
                    if attr_type == bool:
                        setattr(self, line_info[0], attr_type(int(line_info[2])))
                    else:
                        setattr(self, line_info[0], attr_type(line_info[2]))

        self.saving = True
        self.saving_path = path
        self.__init__()

    def save(self, path):

        with open(join(path, 'parameters.txt'), "w") as text_file:

            text_file.write('# -----------------------------------#\n')
            text_file.write('# Parameters of the training session #\n')
            text_file.write('# -----------------------------------#\n\n')

            # Input parameters
            text_file.write('# Input parameters\n')
            text_file.write('# ****************\n\n')
            text_file.write('dataset = {:s}\n'.format(self.dataset))
            text_file.write('network_model = {:s}\n'.format(self.network_model))
            if type(self.num_classes) is list:
                text_file.write('num_classes =')
                for n in self.num_classes:
                    text_file.write(' {:d}'.format(n))
                text_file.write('\n')
            else:
                text_file.write('num_classes = {:d}\n'.format(self.num_classes))
            text_file.write('in_points_dim = {:d}\n'.format(self.in_points_dim))
            text_file.write('in_features_dim = {:d}\n'.format(self.in_features_dim))
            text_file.write('in_radius = {:.3f}\n'.format(self.in_radius))
            text_file.write('input_threads = {:d}\n\n'.format(self.input_threads))

            # Model parameters
            text_file.write('# Model parameters\n')
            text_file.write('# ****************\n\n')

            text_file.write('architecture =')
            for a in self.architecture:
                text_file.write(' {:s}'.format(a))
            text_file.write('\n')
            text_file.write('num_layers = {:d}\n'.format(self.num_layers))
            text_file.write('first_features_dim = {:d}\n'.format(self.first_features_dim))
            text_file.write('use_batch_norm = {:d}\n'.format(int(self.use_batch_norm)))
            text_file.write('batch_norm_momentum = {:.3f}\n\n'.format(self.batch_norm_momentum))
            text_file.write('segmentation_ratio = {:.3f}\n\n'.format(self.segmentation_ratio))

            # KPConv parameters
            text_file.write('# KPConv parameters\n')
            text_file.write('# *****************\n\n')

            text_file.write('first_subsampling_dl = {:.3f}\n'.format(self.first_subsampling_dl))
            text_file.write('num_kernel_points = {:d}\n'.format(self.num_kernel_points))
            text_file.write('density_parameter = {:.3f}\n'.format(self.density_parameter))
            text_file.write('fixed_kernel_points = {:s}\n'.format(self.fixed_kernel_points))
            text_file.write('KP_extent = {:.3f}\n'.format(self.KP_extent))
            text_file.write('KP_influence = {:s}\n'.format(self.KP_influence))
            text_file.write('convolution_mode = {:s}\n'.format(self.convolution_mode))
            text_file.write('trainable_positions = {:d}\n\n'.format(int(self.trainable_positions)))
            text_file.write('modulated = {:d}\n\n'.format(int(self.modulated)))

            # Training parameters
            text_file.write('# Training parameters\n')
            text_file.write('# *******************\n\n')

            text_file.write('learning_rate = {:f}\n'.format(self.learning_rate))
            text_file.write('momentum = {:f}\n'.format(self.momentum))
            text_file.write('lr_decay_epochs =')
            for e, d in self.lr_decays.items():
                text_file.write(' {:d}:{:f}'.format(e, d))
            text_file.write('\n')
            text_file.write('grad_clip_norm = {:f}\n\n'.format(self.grad_clip_norm))


            text_file.write('augment_symmetries =')
            for a in self.augment_symmetries:
                text_file.write(' {:d}'.format(int(a)))
            text_file.write('\n')
            text_file.write('augment_rotation = {:d}\n'.format(self.augment_rotation))
            text_file.write('augment_noise = {:f}\n'.format(self.augment_noise))
            text_file.write('augment_occlusion = {:s}\n'.format(self.augment_occlusion))
            text_file.write('augment_occlusion_ratio = {:.3f}\n'.format(self.augment_occlusion_ratio))
            text_file.write('augment_occlusion_num = {:d}\n'.format(self.augment_occlusion_num))
            text_file.write('augment_scale_anisotropic = {:d}\n'.format(int(self.augment_scale_anisotropic)))
            text_file.write('augment_scale_min = {:.3f}\n'.format(self.augment_scale_min))
            text_file.write('augment_scale_max = {:.3f}\n'.format(self.augment_scale_max))
            text_file.write('augment_color = {:.3f}\n\n'.format(self.augment_color))

            text_file.write('weights_decay = {:f}\n'.format(self.weights_decay))
            text_file.write('gaussian_decay = {:f}\n'.format(self.gaussian_decay))
            text_file.write('batch_averaged_loss = {:d}\n'.format(int(self.batch_averaged_loss)))
            text_file.write('offsets_loss = {:s}\n'.format(self.offsets_loss))
            text_file.write('offsets_decay = {:f}\n'.format(self.offsets_decay))
            text_file.write('batch_num = {:d}\n'.format(self.batch_num))
            text_file.write('max_epoch = {:d}\n'.format(self.max_epoch))
            if self.epoch_steps is None:
                text_file.write('epoch_steps = None\n')
            else:
                text_file.write('epoch_steps = {:d}\n'.format(self.epoch_steps))
            text_file.write('validation_size = {:d}\n'.format(self.validation_size))
            text_file.write('snapshot_gap = {:d}\n'.format(self.snapshot_gap))



